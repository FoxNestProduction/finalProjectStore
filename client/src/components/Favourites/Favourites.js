import React from 'react';
import { useNavigate } from 'react-router';
import { useSelector, shallowEqual } from 'react-redux';
import useMediaQuery from '@mui/material/useMediaQuery';
import Container from '@mui/material/Container';
import Typography from '@mui/material/Typography';
import Stack from '@mui/material/Stack';
import Button from '@mui/material/Button';
import Box from '@mui/material/Box';
import LinearProgress from '@mui/material/LinearProgress';
import FavouriteItem from '../FavouriteItem/FavouriteItem';
import ListItems from '../ListItems/ListItem';
import ProductCardItem from '../ProductCardItem/ProductCardItem';
import buttonBackToMenu from './styles';

const Favourites = () => {
  const navigate = useNavigate();
  const isLgTablet = useMediaQuery('(min-width: 690px)');

  const handlMenuClick = () => {
    navigate('/menu');
  };

  // eslint-disable-next-line no-underscore-dangle
  const favouritesList = useSelector((state) => state.favourites.favourites, shallowEqual);
  const loading = useSelector((state) => state.favourites.loading);
  const error = useSelector((state) => state.favourites.error);

  return (
    <Box position="relative">
      {favouritesList.length !== 0 ? (
        <Box sx={{ mb: '300px', pt: { mobile: '40px', lgTablet: '60px' } }}>
          { isLgTablet && favouritesList
          && (
          <Container sx={{ backgroundColor: 'background.default' }}>
            <Typography variant="h2" component="h3" sx={{ textAlign: 'center', mb: { lgTablet: '34px', desktop: '28px' } }}>Favourite</Typography>
            <Stack direction="column" spacing={3}>
              { favouritesList.map((item) => (
                <Box key={item._id}>
                  <FavouriteItem product={item} />
                </Box>
              ))}
            </Stack>
          </Container>
          )}
          { !isLgTablet && favouritesList
          && (
          <ListItems
            title="Favourite"
            items={favouritesList}
            itemComponent={ProductCardItem}
            actions={null}
          />
          )}
        </Box>
      ) : (
        <Container>
          <Stack direction="column" sx={{ justifyContent: 'center', alignItems: 'center', gap: 5, my: 10 }}>
            <Typography variant="h2" color="primary.main" sx={{ textAlign: 'center' }}>Oops! your wishlist is empty</Typography>
            <Typography variant="h3" color="text.secondary">Fill it with dishes</Typography>
            <Button
              variant="contained"
              onClick={handlMenuClick}
              sx={buttonBackToMenu}
            >
              Back to menu
            </Button>
          </Stack>
        </Container>
      )}
      {loading
      && (
        <Box sx={{ width: '100%', position: 'absolute', top: 0 }}>
          <LinearProgress />
        </Box>
      )}
      {error && <div>{error.statusText}</div>}
    </Box>
  );
};

export default Favourites;
