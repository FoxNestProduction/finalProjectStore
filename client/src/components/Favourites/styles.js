const buttonBackToMenu = {
  height: {
    mobile: '41px',
    tablet: '47px',
    desktop: '60px',
  },
  width: {
    mobile: '130px',
    desktop: '208px',
  },
  fontSize: {
    mobile: '14px',
    desktop: '24px',
  },
  fontWeight: 500,
  transition: 'background-color 0.3s ease, box-shadow 0.3s ease, color 0.3s ease',
  '&:hover': {
    backgroundColor: 'primary.hover',
    boxShadow: '0px 4px 6px rgba(0, 0, 0, 0.3)',
    color: 'text.primaryLight',
  },
  '&:active': {
    backgroundColor: 'common.white',
    boxShadow: '0px 4px 6px rgba(0, 0, 0, 0.3)',
    color: 'primary.main',
    border: '1px solid',
  },
};

export default buttonBackToMenu;
