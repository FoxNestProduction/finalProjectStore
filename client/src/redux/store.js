import { configureStore, getDefaultMiddleware, combineReducers } from '@reduxjs/toolkit';
// todo: правила для redux-persist
// eslint-disable-next-line import/no-extraneous-dependencies
import storage from 'redux-persist/lib/storage';
// eslint-disable-next-line import/no-extraneous-dependencies
import { persistReducer } from 'redux-persist';
import modalSlice from './slices/modalSlice';
import productsSlice from './slices/productsSlice';
import authorizationSlice from './slices/authorizationSlice';
import userSlice from './slices/userSlice';
import partnersSlice from './slices/partnersSlice';
import searchSlice from './slices/searchSlice';
import errorSlice from './slices/errorSlice';
import reviewsSlice from './slices/reviewsSlice';
import favouriteSlice from './slices/favouriteSlice';

const authPersistConfig = {
  key: 'authorization',
  version: 1,
  storage,
};

const userPersistConfig = {
  key: 'user',
  version: 1,
  storage,
};

const searchPersistConfig = {
  key: 'search',
  version: 1,
  storage,
};
const favouritePersistConfig = {
  key: 'favorites',
  version: 1,
  storage,
};

/*
    todo: - ЗВЕРНІТЬ УВАГУ!
      якщо вам потрібно щоб ваш стейт не зникав після перезавантаження
      то юзайте підхід як authorization і user,
      якщо ж ні то варіант написання як для modal всередині reducer
      Якщо що пишіть @Ihor_Kacher
*/
const reducer = combineReducers({
  error: errorSlice,
  modal: modalSlice,
  partners: partnersSlice,
  authorization: persistReducer(authPersistConfig, authorizationSlice),
  user: persistReducer(userPersistConfig, userSlice),
  products: productsSlice,
  search: persistReducer(searchPersistConfig, searchSlice),
  reviews: reviewsSlice,
  favourites: persistReducer(favouritePersistConfig, favouriteSlice),
});

const middleware = getDefaultMiddleware({
  serializableCheck: false,
});

const store = configureStore({
  reducer,
  middleware,
});

export default store;
