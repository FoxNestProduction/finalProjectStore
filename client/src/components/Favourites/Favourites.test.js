import React from 'react';
// import { MemoryRouter, Route, Routes, useLocation, useNavigate  } from 'react-router-dom';
import { MemoryRouter } from 'react-router-dom';
import { Route, Routes, useLocation, useNavigate } from 'react-router';
import { Provider } from 'react-redux';
import { screen, render, fireEvent, waitFor } from '@testing-library/react';
import { useMediaQuery } from '@mui/material';
import store from '../../redux/store';
import Favourites from './Favourites';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
}));
jest.mock('react-router', () => ({
  ...jest.requireActual('react-router'),
  useLocation: jest.fn(() => ({ pathname: '/favourites' })),
  useNavigate: jest.fn(),
  Routes: ({ children }) => children,
  Route: ({ element }) => element,
}));;
jest.mock('@mui/material/', () => ({
  ...jest.requireActual('@mui/material/'),
  useMediaQuery: jest.fn(),
}));

describe('Snapshot test', () => {

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should Favourites render, when arr is empty', () => {
    const navigateMock = jest.fn();
    useNavigate.mockReturnValue(navigateMock);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce([]);

    const { asFragment } = render(
      <MemoryRouter initialEntries={['/favourites']}>
        <Routes>
          <Route path="/favourites" element={<Favourites />} />
        </Routes>
      </MemoryRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.queryByRole('list')).toBeNull();

    const button = screen.getByText('Back to menu');

    fireEvent.click(button);
    expect(navigateMock).toHaveBeenCalledWith('/menu');
  });

  test('should render FavouriteItem', () => {
    const favouritesList = [
      { _id: '1', name: 'pasta', currentPrice: 10.99, imageUrl: 'image1.jpg' },
      { _id: '2', name: 'pizza', currentPrice: 15.99, imageUrl: 'image2.jpg' },
    ];

    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(favouritesList);

    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Favourites />
        </MemoryRouter>
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();

    favouritesList.forEach((item) => {
      const itemName = screen.getByText(item.name);
      const itemPrice = screen.getByText(`$${item.currentPrice.toFixed(2)}`);

      expect(itemName).toBeInTheDocument();
      expect(itemPrice).toBeInTheDocument();
    });
  });
});
