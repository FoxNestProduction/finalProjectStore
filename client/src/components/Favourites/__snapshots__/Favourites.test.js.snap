// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test should Favourites render, when arr is empty 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-79elbk"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
    >
      <div
        class="MuiStack-root css-1idwvdq-MuiStack-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 css-1ni6hz-MuiTypography-root"
        >
          Oops! your wishlist is empty
        </h2>
        <h3
          class="MuiTypography-root MuiTypography-h3 css-rw3qn6-MuiTypography-root"
        >
          Fill it with dishes
        </h3>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-9vr5c8-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Back to menu
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
